### PR TITLE
Add child product channel sync status visibility at parent level

### PR DESCRIPTION
**Pull request by ncounter**
## What does this PR change?

In the Product page a children product channel sync fails, we need to report to the collapsed parent tree this information.
No reason to keep the `progress bar` percentage indicator as it was not really reporting a sync progress, but just a percentage of *how many channels finished on how many channels it is syncing*

@hustodemon [this is the commit](https://github.com/uyuni-project/uyuni/pull/2020/commits/048de8f77fe0a57a54fd10e2eaeba440a346e7ef) where I really need a review, because I am not sure about performances and if it is the correct way of doing it. The scope of that commit is just to add the `channelId` to each channel of products when the channel exists. I am not sure if it is better to do as I did or just iterating on the complete tree at the very end just once and add finally.

## GUI diff

Before:
Screenshot from 2020-03-16 14-54-54

After:
Screenshot from 2020-03-20 23-36-44

Screenshot from 2020-03-20 22-42-42

- [x] **DONE**

## Documentation
- No documentation needed: @Loquacity do you think we need to document this somehow? Screenshots?

- [x] **DONE**

## Test coverage
- No tests: @srbarrios ?

- [x] **DONE**

## Links

Fixes https://github.com/SUSE/spacewalk/issues/9884
Tracks # 

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"		 
- [ ] Re-run test "java_pgsql_tests"		 
- [ ] Re-run test "ruby_rubocop"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"		 
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"		 
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by ncounter** [comment]
@mcalmer do you think it's ok if we *drop* the `re-sync` product button in the product page? I think it is really something should stay in the channel section. Here now we have a clear status of the sync, we can open the exact product which has the problem, but scheduling a channel sync from the product page is something I'd really like to avoid. It's a channel action and it should be taken only from a channel administrator. What do you think?

**Comment by mcalmer** [comment]
> @mcalmer do you think it's ok if we _drop_ the `re-sync` product button in the product page? I think it is really something should stay in the channel section. Here now we have a clear status of the sync, we can open the exact product which has the problem, but scheduling a channel sync from the product page is something I'd really like to avoid. It's a channel action and it should be taken only from a channel administrator. What do you think?

The problem is, that we talk about `Vendor Channels`. And you cannot execute a sync on it on the channel page.
So we would loose the possibility to re-sync and this is not ok.

**Comment by Loquacity** [comment]
> * No documentation needed: @Loquacity do you think we need to document this somehow? Screenshots?

Probably in the Ref Guide

**Comment by ncounter** [comment]
> So we would loose the possibility to re-sync and this is not ok.

Ack. As discussed offline I readded the resync button for the entire product (with some hidden improvements/bugfixes actually). For the scope of this PR I think it is ok to keep it as it is. Additional improvements and changes can wait for later (e.g. adding sync button to the channel detail page, adding buttons to the channels list popup of the product, etc).

File: java/code/src/com/suse/manager/webui/controllers/ProductsController.java [hunk 1]
```diff
@@ -319,6 +318,10 @@ public static String data(Request request, Response response, User user) {
                                             s.getStatus(),
                                             s.getChannels().stream().map(c ->
                                                     new ChannelJson(
+                                                            s.getStatus().equals(MgrSyncStatus.INSTALLED) &&
+                                                                    ChannelFactory.lookupByLabel(c.getLabel()) != null ?
+                                                                    ChannelFactory.lookupByLabel(c.getLabel()).getId() :
+                                                                    -1L,
```
Comment (mcalmer) [inline java/code/src/com/suse/manager/webui/controllers/ProductsController.java]: I think this is too expensive and there is a better way to get the ID.

Check line 300: there is a map "channelByLabel" of all Channels. Please use that map to get the Channel and the ID.

File: java/code/src/com/suse/manager/webui/controllers/ProductsController.java [hunk 2]
```diff
@@ -365,6 +368,10 @@ public static String data(Request request, Response response, User user) {
                         rootExtensions,
                         syncProduct.getChannels().stream().map(c ->
                                 new ChannelJson(
+                                        syncProduct.getStatus().equals(MgrSyncStatus.INSTALLED) &&
+                                                ChannelFactory.lookupByLabel(c.getLabel()) != null ?
+                                                ChannelFactory.lookupByLabel(c.getLabel()).getId() :
+                                                -1L,
```
Comment (mcalmer) [inline java/code/src/com/suse/manager/webui/controllers/ProductsController.java]: Same as above

File: java/code/src/com/suse/manager/webui/controllers/ProductsController.java (hunk 1 above)
Comment (ncounter) [reply java/code/src/com/suse/manager/webui/controllers/ProductsController.java]: Thanks for the hint. Fixed

**Comment by Loquacity** [comment]
> * No documentation needed: @Loquacity do you think we need to document this somehow? Screenshots?

No, I don't think so.

File: java/code/src/com/suse/manager/webui/controllers/ProductsController.java [hunk 3]
```diff
@@ -319,6 +318,10 @@ public static String data(Request request, Response response, User user) {
                                             s.getStatus(),
                                             s.getChannels().stream().map(c ->
                                                     new ChannelJson(
+                                                            Opt.fold(Optional.ofNullable(
+                                                                            channelByLabel.get(c.getLabel())),
+                                                                    () -> -1L,
+                                                                    x -> x.getId()),
```
Comment (hustodemon) [inline java/code/src/com/suse/manager/webui/controllers/ProductsController.java]: Nitpick: (this is **totally** matter of taste, so disregard it, if you don't like it.)

I found an alternative syntax to be more readable:

```java
Optional.ofNullable(channelByLabel.get(c.getLabel()))
    .map(chan -> chan.getId())
    .orElse(-1L)
```

File: java/code/src/com/suse/manager/webui/controllers/ProductsController.java [hunk 4]
```diff
@@ -365,6 +368,9 @@ public static String data(Request request, Response response, User user) {
                         rootExtensions,
                         syncProduct.getChannels().stream().map(c ->
                                 new ChannelJson(
+                                        Opt.fold(Optional.ofNullable(channelByLabel.get(c.getLabel())),
+                                                () -> -1L,
+                                                x -> x.getId()),
```
Comment (hustodemon) [inline java/code/src/com/suse/manager/webui/controllers/ProductsController.java]: Same here, feel free not to agree :wink:

File: java/code/src/com/suse/manager/webui/controllers/ProductsController.java (hunk 3 above)
Comment (ncounter) [reply java/code/src/com/suse/manager/webui/controllers/ProductsController.java]: I am always undecided which is better actually.
`Opt.fold` is more complex but `Optional.ofNullable` uses `.map()` and I don't like to map when I know it is just a single object that I have to work with, it is less intuitive in such a case. So either one way or the other in this case has pros and cons to me. And if you say the second one is clearer, I'm fine with it and I trust you more in the Java aspect. Changing accordingly :+1:

File: web/html/src/manager/admin/setup/products/products.js [hunk 5]
```diff
@@ -735,49 +745,101 @@ class CheckListItem extends React.Component {
...
+          <span className="text-info" title={t('Child products channels in progress')}>
+            (<i className="fa fa-spinner fa-spin"></i>)
+          </span>;
+      }
+      else if (childProductChannels.filter(c => c.status == _CHANNEL_STATUS.synced).length > 0) {
+        childProductChannelSyncContent =
+          <span className="text-success" title={t('Child products channels synced')}>
+            (<i className="fa fa-check-circle"></i>)
+          </span>;
       }
```
[the remark this answers is not in the document]
Comment (ncounter) [reply web/html/src/manager/admin/setup/products/products.js]: Yeah, it is not exactly the same code, but with some small extracted differences I was able to make a single function to reuse the shared part. :+1:

**Pull request merged**